src/delta.c

```diff
@@ -38,86 +38,164 @@
+	if (xattrs_compare(prefixed_from_file, to_staged) != 0) {
+		fprintf(stderr, "xattrs comparison failed\n");
+		goto remove_and_fail;
 	}
-	xattrs_copy(origin, filename);
 
-	if (!verify_file(file, filename)) {
-		unlink_all_staged_content(file);
-		goto out;
+	char hash[SWUPD_HASH_LEN];
+	if (!compute_hash_from_file(to_staged, &hash[0])) {
+		fprintf(stderr, "hash calculation failed\n");
+		goto remove_and_fail;
 	}
+	if (!hash_equal(hash, to_hash)) {
+		fprintf(stderr, "hash mismatch\n");
+		goto remove_and_fail;
+	}
+
+	free_string(&prefixed_from_file);
+	return true;
+
+remove_and_fail:
+	(void)remove(to_staged);
+fail:
+	free_string(&prefixed_from_file);
 
-	if (xattrs_compare(origin, filename) != 0) {
-		unlink_all_staged_content(file);
-		goto out;
+	return false;
+}
+
+void apply_deltas(struct manifest *current_manifest)
+{
+	char *delta_dir;
+	string_or_die(&delta_dir, "%s/delta", state_dir);
+
+	DIR *dir = opendir(delta_dir);
+	if (!dir) {
+		/* No deltas available to apply. */
+		return;
 	}
 
-	unlink(deltafile);
+	struct dirent *ent;
+	while ((ent = readdir(dir))) {
+		if (!strcmp(ent->d_name, ".") || !strcmp(ent->d_name, "..")) {
+			continue;
+		}
+
+		/* Delta files have the form VERSION-VERSION-FROM_HASH-TO_HASH, we need to
+		 * extract FROM and TO. */
+		char *delta_name = ent->d_name;
+		char *s = delta_name;
+
+		s = index(s, '-');
+		if (!s) {
+			fprintf(stderr, "Skipping invalid delta file %s/%s\n", delta_dir, delta_name);
+			continue;
+		}
+		s++;
+
+		s = index(s, '-');
+		if (!s) {
+			fprintf(stderr, "Skipping invalid delta file %s/%s\n", delta_dir, delta_name);
+			continue;
+		}
+		s++;
+
+		/* Note: SWUPD_HASH_LEN accounts for the nul-terminator after the hash. */
+
+		/* Two hashes (without zero) plus one dash in the middle. */
+		if (strlen(s) != ((2 * (SWUPD_HASH_LEN - 1)) + 1)) {
+			fprintf(stderr, "Skipping invalid delta file %s/%s\n", delta_dir, delta_name);
+			continue;
+		}
+
+		if (s[SWUPD_HASH_LEN - 1] != '-') {
+			fprintf(stderr, "Skipping invalid delta file %s/%s\n", delta_dir, delta_name);
+			continue;
+		}
+
+		char from_buf[SWUPD_HASH_LEN] = {};
+		memcpy(from_buf, s, SWUPD_HASH_LEN - 1);
+
+		char *from = &from_buf[0];
+		char *to = &s[SWUPD_HASH_LEN];
+
+		char *to_staged;
+		string_or_die(&to_staged, "%s/staged/%s", state_dir, to);
+
+		struct stat stat;
+		if (lstat(to_staged, &stat) == 0) {
+			free_string(&to_staged);
+			continue;
+		}
+
+		/* TODO: Sort the list of files by hash then walk it in parallel with readdir. */
+
+		struct list *ll = list_head(current_manifest->files);
+
+		struct file *found = NULL;
+		while (ll) {
+			struct file *file = ll->data;
+			ll = ll->next;
+
+			if (!file->is_deleted && !file->is_ghosted && file->is_file && hash_equal(file->hash, from)) {
+				found = file;
+				break;
+			}
+		}
+
+		if (!found) {
+			fprintf(stderr, "Skipping delta file %s/%s because the 'from' file for delta was found\n", delta_dir, delta_name);
```

Comment:
You mean the 'from' file for delta was **not** found?

Reply:
Done.